### PR TITLE
Decompression Implement Base on OpenMP

### PR DESCRIPTION
**Pull request by GaiusCjl**

### CODE REVIEW

File: hashmap.c [hunk 1]
```diff
@@ -0,0 +1,171 @@
+#include "process.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+// Node structure for the linked list in each hash table entry
```
Comment (JoernZheng) [inline hashmap.c:6]: You can remove unused comments

File: decompression.c [hunk 2]
```diff
@@ -106,24 +226,60 @@ void decompress_zwz(const char *file_path, const char *output_dir_path) {
...
+    dir = opendir(source_dir_path);
+    if (dir == NULL) perror("Could not open source directory");
+    else {
+        // multi-threads decompressing .zwz files
+        #pragma omp parallel 
+        {
+            char* d_name_copy;
+
+            // update ent and d_name_copy
+            #pragma omp critial
```
Comment (JoernZheng) [inline decompression.c:248]: Do you mean `#pragma omp critical` in this context?

File: decompression.c [hunk 3]
```diff
@@ -106,24 +226,60 @@ void decompress_zwz(const char *file_path, const char *output_dir_path) {
...
+            while (!readEnd){
+                if (strstr(d_name_copy, ".zwz") != NULL) {
+                    char file_path[1024];
+                    snprintf(file_path, sizeof(file_path), "%s/%s", source_dir_path, d_name_copy);
+                    // Step 2: Process each .zwz file
+                    decompress_zwz(file_path, output_dir_path);
+                }
+                free(d_name_copy);
+                // update ent and d_name_copy
+                #pragma omp critial
```
Comment (JoernZheng) [inline decompression.c:268]: Same

File: decompression.c [hunk 4]
```diff
@@ -1,5 +1,54 @@
 #include "process.h"
 #include <dirent.h>
+#include <string.h>
```
[the remark this answers is not in the document]
Comment (GaiusCjl) [reply decompression.c]: Thanks, I have removed those header files.

File: .gitignore [hunk 5]
```diff
@@ -64,3 +64,5 @@ dkms.conf
 #CMakeLists.txt
 cmake-build-debug/
 
+tmp/*
```
[the remark this answers is not in the document]
Comment (GaiusCjl) [reply .gitignore]: Thanks, I have removed it.

File: hashmap.c (hunk 1 above)
Comment (GaiusCjl) [reply hashmap.c]: Thanks, I have removed those comments.

File: processSizeFile.c
[the remark this answers is not in the document]
Comment (GaiusCjl) [reply processSizeFile.c:1]: Thanks, I have removed it!

File: decompression.c (hunk 2 above)
Comment (GaiusCjl) [reply decompression.c]: Sorry for the typo, I changed it to #pragma omp critical and tested it. It worked fine.

**Review by JoernZheng — APPROVED** [review]

**Pull request merged**